### PR TITLE
[AAE-1197] The datetime validation error message can be improved to a more user friendly message.

### PR DESCRIPTION
**Pull request by dhrn**
fixes https://issues.alfresco.com/jira/browse/AAE-1197

JSON validations with **custom error message** have been added
**Overview**:-

We can define the custom error message based on the failed validation.
Below, custom error message configured for two cases (pattern mismatch and wrong data type)
```
            "properties": {
              "value": {
                "type": "string",
                "pattern": "<pattern -goes here>",
                "message": {
                  "pattern": "invalid date {{name}} {{id}}",
                  "type": "mismatch type {{name}} {{id}}. expected type is date"
                }
              }
           
```

**Message pattern Explanation:**- 

```invalid date {{name}} {{id}}```

**{{name}} {{id}}** ->  values will replaced with actual **process variable name and id** in runtime

**New Behaviour:-** 
here is the updated error message pattern
**Date:**
```
 "pattern": "Invalid date - {{name}}({{id}})",
 "type": "Mismatch value type - {{name}}({{id}}). Expected type is date"
```
**DateTime**
```
  "pattern": "Invalid datetime - {{name}}({{id}})",
  "type": "Mismatch value type - {{name}}({{id}}). Expected type is datetime"
```

**Integer**
```
 "type": "Mismatch value type - {{name}}({{id}}). Expected type is integer"
```
**Boolean**
```
 "type": "Mismatch value type - {{name}}({{id}}). Expected type is boolean"
```
**Json**
```
 "type": "Mismatch value type - {{name}}({{id}}). Expected type is json"
```
**File**
```
 "type": "Mismatch value type - {{name}}({{id}}). Expected type is file"
```

**Here is the example:-** 
```
Mismatch value type - dateVariable(c297ec88-0ecf-4841-9b0f-2ae814957c68). Expected type is date
Mismatch value type - stringVariable(c297ec88-0ecf-4841-9b0f-2ae814957c68). Expected type is string
Mismatch value type - case4(e0740a3a-fec4-4ee5-bece-61f39df2a47k). Expected type is datetime
Mismatch value type - objectVariable(c297ec88-0ecf-4841-9b0f-2ae814957c68). Expected type is json
Mismatch value type - booleanVariable(c297ec88-0ecf-4841-9b0f-2ae814957c68). Expected type is boolean
Invalid date - dateVariable(c297ec88-0ecf-4841-9b0f-2ae814957c68)
Invalid datetime - case4(e0740a3a-fec4-4ee5-bece-61f39df2a47f)

```

### CODE REVIEW

File: activiti-cloud-services-org/activiti-cloud-services-org-service/src/main/java/org/activiti/cloud/services/organization/validation/JsonSchemaModelValidator.java [hunk 1]
```diff
@@ -81,12 +85,72 @@ public void validate(byte[] bytes,
...
+            message = message.replace("{{type}}", getValueFromJson(path, prcessExtenstionJson));
+        } else if (message.contains("{{name}}")) {
+            path[length] = "name";
+            message = message.replace("{{name}}", getValueFromJson(path, prcessExtenstionJson));
+        } else if (message.contains("{{id}}")) {
+            path[length] = "id";
+            message = message.replace("{{id}}", getValueFromJson(path, prcessExtenstionJson));
+        }
+
+       if (message.contains( "{{name}}") || message.contains( "{{id}}") || message.contains( "{{type}}") || message.contains( "{{value}}"))
```
Comment (erdemedeiros) [inline activiti-cloud-services-org/activiti-cloud-services-org-service/src/main/java/org/activiti/cloud/services/organization/validation/JsonSchemaModelValidator.java]: Please, use brackets around `if` statement body, even if it has a single line.

File: activiti-cloud-services-org/activiti-cloud-services-org-service/src/main/java/org/activiti/cloud/services/organization/validation/JsonSchemaModelValidator.java [hunk 2]
```diff
@@ -81,12 +85,72 @@ public void validate(byte[] bytes,
...
                 .orElse(null);
         return createModelValidationError(validationException.getErrorMessage(),
-                                          validationException.getMessage(),
+                                          description,
                                           schema);
     }
+
+    private String resolveExpression(String message, String  pointerToViolation, JSONObject prcessExtenstionJson) {
+        final String path[] = pointerToViolation.replace("#/", "").split("/");
+        int length = path.length - 1;
```
Comment (erdemedeiros) [inline activiti-cloud-services-org/activiti-cloud-services-org-service/src/main/java/org/activiti/cloud/services/organization/validation/JsonSchemaModelValidator.java]: Should this variable be named `length` or `*position` or `*index`? From what I see it's defining a position that should be overridden.

File: activiti-cloud-services-org/activiti-cloud-services-org-service/src/main/java/org/activiti/cloud/services/organization/validation/JsonSchemaModelValidator.java [hunk 3]
```diff
@@ -81,12 +85,72 @@ public void validate(byte[] bytes,
...
+        }
+
+       if (message.contains( "{{name}}") || message.contains( "{{id}}") || message.contains( "{{type}}") || message.contains( "{{value}}"))
+           return resolveExpression(message, pointerToViolation, prcessExtenstionJson);
+
+        return  message;
+    }
+
+    private String getValueFromJson(String[] path, JSONObject processExtensionJson) {
+        JSONObject parent = null; String value = "";
```
Comment (erdemedeiros) [inline activiti-cloud-services-org/activiti-cloud-services-org-service/src/main/java/org/activiti/cloud/services/organization/validation/JsonSchemaModelValidator.java]: Please, declare one variable per line.

File: activiti-cloud-services-org/activiti-cloud-services-org-service/src/main/java/org/activiti/cloud/services/organization/validation/JsonSchemaModelValidator.java [hunk 4]
```diff
@@ -81,12 +85,72 @@ public void validate(byte[] bytes,
...
+
+        if(path.length > 1) {
+            parent = processExtensionJson.getJSONObject(path[0]);
+
+            for(int iterator=1; iterator < path.length - 1; iterator++)
```
Comment (erdemedeiros) [inline activiti-cloud-services-org/activiti-cloud-services-org-service/src/main/java/org/activiti/cloud/services/organization/validation/JsonSchemaModelValidator.java]: Missing brackets.

File: activiti-cloud-services-org/activiti-cloud-services-org-service/src/main/java/org/activiti/cloud/services/organization/validation/JsonSchemaModelValidator.java [hunk 5]
```diff
@@ -81,12 +85,72 @@ public void validate(byte[] bytes,
...
+                parent = parent.getJSONObject(path[iterator]);
+
+            value = parent.getString(path[path.length - 1]);
+        }
+
+        if (path.length == 1)
```
Comment (erdemedeiros) [inline activiti-cloud-services-org/activiti-cloud-services-org-service/src/main/java/org/activiti/cloud/services/organization/validation/JsonSchemaModelValidator.java]: Missing brackets.

File: activiti-cloud-services-org/activiti-cloud-services-org-service/src/main/resources/schema/process-extensions-schema.json [hunk 6]
```diff
@@ -84,7 +84,11 @@
             "properties": {
               "value": {
                 "type": "string",
-                "pattern": "^[0-9]{4}-(((0[13578]|(10|12))-(0[1-9]|[1-2][0-9]|3[0-1]))|(02-(0[1-9]|[1-2][0-9]))|((0[469]|11)-(0[1-9]|[1-2][0-9]|30)))$"
+                "pattern": "^[0-9]{4}-(((0[13578]|(10|12))-(0[1-9]|[1-2][0-9]|3[0-1]))|(02-(0[1-9]|[1-2][0-9]))|((0[469]|11)-(0[1-9]|[1-2][0-9]|30)))$",
+                "message": {
+                  "pattern": "invalid date {{name}} {{id}}",
+                  "type": "mismatch type {{name}} {{id}}. expected type is date"
```
Comment (erdemedeiros) [inline activiti-cloud-services-org/activiti-cloud-services-org-service/src/main/resources/schema/process-extensions-schema.json]: Would it make sense to start to use capital letters for the first letter of `pattern` and `type`? 
- `Invalid date {{name}} {{id}}`
- `Mismatch type {{name}} {{id}}. Expected type is date.`

File: activiti-cloud-services-org/activiti-cloud-services-org-service/src/main/resources/schema/process-extensions-schema.json [hunk 7]
```diff
@@ -149,7 +164,10 @@
           "then": {
             "properties": {
               "value": {
-                "type": "boolean"
+                "type": "boolean",
+                "message": {
+                  "type": "mismatch type {{type}} - {{name}} {{id}}. expected type is boolean"
```
Comment (erdemedeiros) [inline activiti-cloud-services-org/activiti-cloud-services-org-service/src/main/resources/schema/process-extensions-schema.json]: Some messages for type mismatch contains `{{type}} - {{name}} {{id}` (i.e. `boolean`, `json`, `file`); others contains only `{{name}} {{id}`. Is there any specific reason for that?

**Review by erdemedeiros — CHANGES_REQUESTED** [review]
@dhrn, please check my comments and questions bellow.

File: activiti-cloud-services-org/activiti-cloud-services-org-service/src/main/resources/schema/process-extensions-schema.json (hunk 6 above)
Comment (dhrn) [reply activiti-cloud-services-org/activiti-cloud-services-org-service/src/main/resources/schema/process-extensions-schema.json]: yes. done

File: activiti-cloud-services-org/activiti-cloud-services-org-service/src/main/resources/schema/process-extensions-schema.json (hunk 7 above)
Comment (dhrn) [reply activiti-cloud-services-org/activiti-cloud-services-org-service/src/main/resources/schema/process-extensions-schema.json]: I forgot to update here. solved now. thanks

**Review by erdemedeiros — APPROVED** [review]

**Pull request merged**